File: LLWallPaper.App/Stores/HistoryStore.cs
using System.IO;
using System.Linq;
using System.Text.Json;
using LLWallPaper.App.Models;
using LLWallPaper.App.Utils;

namespace LLWallPaper.App.Stores;

public sealed class HistoryStore
{
    private readonly AppLogger _logger;

    public HistoryStore(AppLogger logger)
    {
        _logger = logger;
    }

    public void Append(HistoryEntry entry, int maxEntries)
    {
        var state = LoadState();
        state.BasePath = ResolveBasePath(state.BasePath);
        state.Entries.Add(entry);
        if (maxEntries > 0 && state.Entries.Count > maxEntries)
        {
            var excess = state.Entries.Count - maxEntries;
            state.Entries.RemoveRange(0, excess);
        }
        SaveState(state);
    }

    public HistoryState GetState()
    {
        return LoadState();
    }

    public IReadOnlyList<HistoryEntry> GetAllEntries()
    {
        return LoadState().Entries;
    }

    public void TrimToMax(int maxEntries)
    {
        if (maxEntries <= 0)
        {
            return;
        }

        var state = LoadState();
        if (state.Entries.Count <= maxEntries)
        {
            return;
        }

        var excess = state.Entries.Count - maxEntries;
        state.Entries.RemoveRange(0, excess);
        SaveState(state);
    }

    public IReadOnlyList<string> GetRecentKeys(int count)
    {
        return GetRecentEntries(count).Select(entry => entry.Key).ToList();
    }

    public IReadOnlyList<HistoryEntry> GetRecentEntries(int count)
    {
        if (count <= 0)
        {
            return Array.Empty<HistoryEntry>();
        }

        var entries = LoadState().Entries;
        if (entries.Count == 0)
        {
            return Array.Empty<HistoryEntry>();
        }

        var skip = Math.Max(0, entries.Count - count);
        return entries.Skip(skip).ToList();
    }

    public IReadOnlyList<string> GetRecentLocalPaths(int count)
    {
        if (count <= 0)
        {
            return Array.Empty<string>();
        }

        var entries = LoadState().Entries;
        if (entries.Count == 0)
        {
            return Array.Empty<string>();
        }

        var skip = Math.Max(0, entries.Count - count);
        return entries
            .Skip(skip)
            .Select(entry => AppPaths.GetCachePathForKey(entry.Key))
            .Where(path => !string.IsNullOrWhiteSpace(path) && File.Exists(path))
            .ToList();
    }

    private HistoryState LoadState()
    {
        AppPaths.EnsureDirectories();
        if (File.Exists(AppPaths.HistoryPath))
        {
            try
            {
                var json = File.ReadAllText(AppPaths.HistoryPath);
                var state =
                    JsonSerializer.Deserialize<HistoryState>(json, JsonOptions.Default)
                    ?? new HistoryState();
                state.BasePath = ResolveBasePath(state.BasePath);
                state.Entries ??= new List<HistoryEntry>();
                return state;
            }
            catch (Exception ex)
            {
                _logger.Error("Failed to read history entries.", ex);
                return new HistoryState { BasePath = ResolveBasePath(string.Empty) };
            }
        }

        var empty = new HistoryState { BasePath = ResolveBasePath(string.Empty) };
        SaveState(empty);
        return empty;
    }

    private void SaveState(HistoryState state)
    {
        AppPaths.EnsureDirectories();
        state.BasePath = ResolveBasePath(state.BasePath);
        var json = JsonSerializer.Serialize(state, JsonOptions.Default);
        File.WriteAllText(AppPaths.HistoryPath, json);
    }

    private static string ResolveBasePath(string basePath)
    {
        if (!string.IsNullOrWhiteSpace(basePath))
        {
            return basePath;
        }

        return AppPaths.CacheRoot;
    }
}

File: LLWallPaper.Mac/Sources/LLWallPaperMacCore/Infrastructure/HistoryStore.swift
import Foundation

public final class HistoryStore: HistoryRecording {
  private let logger: AppLogger

  public init(logger: AppLogger) {
    self.logger = logger
  }

  public func append(_ entry: HistoryEntry, maxEntries: Int) {
    var state = loadState()
    state.basePath = resolvedBasePath(state.basePath)
    state.entries.append(entry)
    if maxEntries > 0 && state.entries.count > maxEntries {
      state.entries.removeFirst(state.entries.count - maxEntries)
    }
    saveState(state)
  }

  public func getState() -> HistoryState {
    loadState()
  }

  public func trimToMax(_ maxEntries: Int) {
    guard maxEntries > 0 else {
      return
    }

    var state = loadState()
    guard state.entries.count > maxEntries else {
      return
    }

    state.entries.removeFirst(state.entries.count - maxEntries)
    saveState(state)
  }

  public func recentKeys(count: Int) -> Set<String> {
    Set(recentEntries(count: count).map(\.key))
  }

  public func recentLocalUrls(count: Int) -> Set<URL> {
    Set(
      recentEntries(count: count)
        .compactMap { AppPaths.cacheUrl(for: $0.key) }
        .filter { FileManager.default.fileExists(atPath: $0.path) }
    )
  }

  private func recentEntries(count: Int) -> [HistoryEntry] {
    guard count > 0 else {
      return []
    }

    let entries = loadState().entries
    guard !entries.isEmpty else {
      return []
    }

    return Array(entries.suffix(count))
  }

  private func loadState() -> HistoryState {
    do {
      try AppPaths.ensureDirectories()
      guard FileManager.default.fileExists(atPath: AppPaths.historyUrl.path) else {
        let state = HistoryState(basePath: resolvedBasePath(""))
        saveState(state)
        return state
      }

      let data = try Data(contentsOf: AppPaths.historyUrl)
      var state = try JsonCoding.decoder.decode(HistoryState.self, from: data)
      state.basePath = resolvedBasePath(state.basePath)
      return state
    } catch {
      logger.error("Failed to read history entries.", error)
      return HistoryState(basePath: resolvedBasePath(""))
    }
  }

  private func saveState(_ state: HistoryState) {
    do {
      try AppPaths.ensureDirectories()
      var state = state
      state.basePath = resolvedBasePath(state.basePath)
      let data = try JsonCoding.encoder.encode(state)
      try data.write(to: AppPaths.historyUrl, options: .atomic)
    } catch {
      logger.error("Failed to save history entries.", error)
    }
  }

  private func resolvedBasePath(_ basePath: String) -> String {
    if !basePath.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
      return basePath
    }

    return AppPaths.cacheRoot.path
  }
}
